File: Top-Interview-Questions/Array/Medium-Collection/48-Rotate_Image.cpp
// Time Complexity: O(n^n)
// Space Complexity: O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 7.4 MB, less than 100.00%

// Best Case == Worst Case
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        
        
        int a,b;
        
        a = 0;
        b = matrix.size()-1;
        
        while(a<b){
            for(int i = 0; i < b-a; i++){
                
                swap(matrix[a][a+i], matrix[a+i][b]);
                swap(matrix[a][a+i], matrix[b][b-i]);
                swap(matrix[a][a+i], matrix[b-i][a]);
            }
            a++;
            b--;
        }
        
        
    }
};

File: Top-Interview-Questions/Array/Easy-Collection/48-rotate-image.py
class Solution:
    # Sol 1. Swapping Matrix Element
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        
        for i in range(n):
            matrix[i].reverse()
        # Time Complexity: O(n^2)
        # Space Complexity: O(1)
        # Run Code Runtime: 44 ms
        # Submit Runtime: 56 ms, 10.18% faster.
        # Submit Memory: 13.9 mb, 6.25% less.
